Fixed checking if `append_entries_request` batches are already present in follower log

mmaslankaprv opened the pull request:
### Background
When follower receives an append entries request that `prev_log_index` is smaller than its own `prev_log_index` it validates if the batches from the request matches (by checking a batch offset and corresponding term) its own log. It that is the case the batches are skipped to prevent truncation of valid batches and avoid the data loss.


#### Negative `append_entries_request::prev_log_index`

The validation of already matching batches was broken if they happened to be at the beginning of the log. In this case the `prev_log_index` is not initialised. This case was not correctly handled by the logic calculating the next offset when checking matching batches.

#### Replying with success when all request batches match

When follower receives an append entries request with the vector of records that are all present in its own log and their offsets and terms match it should reply with success and correct `last_dirty_log_index`.
This way a leader instead of moving the follower `next_offset` backward can start recovery process and deliver batches which the follower is missing.

<!--
See https://github.com/redpanda-data/redpanda/blob/dev/CONTRIBUTING.md#pull-request-body
for more details and examples of what is expected in a PR body.

Content in this top section is REQUIRED. Describe, in plain language, the motivation
behind the change (bug fix, feature, improvement) in this PR and how the included
commits address it.

Add the GitHub keyword `Fixes` to link to bug(s) this PR will fix, e.g.
  Fixes #ISSUE-NUMBER, Fixes #ISSUE-NUMBER, ...

If this PR is a backport, link to the original with `Backport of PR`, e.g.
  Backport of PR #PR-NUMBER
-->

## Backports Required

<!-- Checking at least one of the checkboxes is REQUIRED if this PR is not a backport. -->

- [ ] none - not a bug fix
- [ ] none - this is a backport
- [ ] none - issue does not exist in previous branches
- [ ] none - papercut/not impactful enough to backport
- [x] v24.3.x
- [x] v24.2.x
- [x] v24.1.x

## Release Notes

<!--
If the changes in this PR do not need to be mentioned in the release
notes, then don't add a sub-section and simply list `none`, e.g.

* none

Otherwise, adding a sub-section or `none` is REQUIRED if the PR is not a backport PR.
If this is a backport PR, adding contents to this section will override
the release notes section inherited from the original PR to dev.

Add one or more of the sub-sections with a short description bullet
point of the change, e.g.

### Bug Fixes

* Short description of the bug fix if this is a PR to `dev` branch.

### Features

* Short description of the feature. Explain how to configure.

### Improvements

* Short description of how this PR improves existing behavior.

-->
### Bug Fixes
- fixes a very rare situation in which Raft leader can enter into infinite loop trying to recover follower.

mmaslankaprv committed:
r/consensus: fixed validating already matching batches last offset

When follower receives an append entries request that `prev_log_index` is
smaller than its own `prev_log_index` it validates if the
batches from the request matches (by checking a batch offset and
corresponding term) its own log. It that is the case the batches are
skipped to prevent truncation of valid batches and avoid the data loss.

The validation of already matching batches was broken if they happened
to be at the beginning of the log. In this case the `prev_log_index` is
not initialized being negative. This case was not correctly handled by
the logic calculating the next offset when checking matching batches.

That lead to a situation in which a range of batches starting with 0 was
never matching.

Fixed the issue by correctly adjusting the `prev_log_index` if it is
uninitialized.

Signed-off-by: Michał Maślanka <michal@redpanda.com>

mmaslankaprv committed:
r/consensus: do not reply with failure if follower batches match

When follower receives an append entries request with the vector of
records that are all present in its own log and their offsets and terms
match it should reply with success and correct `last_dirty_log_index`.
This way a leader instead of moving the follower `next_offset` backward
can start recovery process and deliver batches which the follower is
missing.

Signed-off-by: Michał Maślanka <michal@redpanda.com>

bashtanov commented on src/v/raft/consensus.cc:
```diff
@@ -2079,6 +2079,25 @@ consensus::do_append_entries(append_entries_request&& r) {
+
+                reply.last_dirty_log_index = adjusted_prev_log_index;
+                // limit the last flushed offset as the adjusted_prev_log_index
+                // may have not yet been flushed.
```
Does it mean `_flushed_offset` may store an offset that has not been flushed? In this case, could you add a comment to `_flushed_offset` to explain what it actually denotes?

mmaslankaprv replied:
the `adjusted_prev_log_index` may be greater than `_flushed_offset`

bashtanov replied:
yeah, but why can't we reply with `_flushed_offset` if it is larger than `adjusted_prev_log_index`?

mmaslankaprv replied:
we do not want leader to see flushed offset which is larger then last log offset.

bashtanov replied:
so is `reply.last_flushed_log_index` the latest flushed log index that matches with the leader's log?

mmaslankaprv replied:
no, not really. This check is to hold an invariant of `flushed_offset <= log_end_offset`

bashtanov replied:
What is `log_end_offset` here? Is it a field of any structure?

bharathv replied:
>    reply.last_flushed_log_index = std::min(adjusted_prev_log_index, _flushed_offset);

does this run a risk of `last_flushed_log_index`  moving backwards, as seen by the leader?  (which I think will have implications on leader commit index computation?

Perhaps we should just set  

> reply.last_flushed_log_index = _flushed_offset;

and let the _flushed_offset computation be monotonic

mmaslankaprv replied:
This discussion was sorted out in the conversion

vbotbuildovich commented:
<!--TEST RESULTS-->
<!--750ff1c2-f48d-4653-884f-a7af156cb987-->
<!--TEST RESULTS-->
<!--30044cda-3a18-441b-bf08-0190229a8ffb-->
<!--TEST RESULTS-->
<!--d12be7c0-7330-4670-9dfa-8fd42847f405-->
<!--title-->
## CI test results
<!--/title-->
<!--body-->
<details>
<summary>test results on <a href="https://buildkite.com/redpanda/redpanda/builds/61563">build#61563</a></summary>


| test_id | test_kind | job_url | test_status | passed |
|:------------------------------------------------------------------------------------------------------------------------------------------------------------|:------------|:------------------------------------------------------------------------------------------|:--------------|:---------|
| rptest.tests.compaction_recovery_test.CompactionRecoveryTest.test_index_recovery | ducktape | https://buildkite.com/redpanda/redpanda/builds/61563#0194d20c-ff99-4b94-b2f7-a64d44ed7679 | FLAKY | 1/3 |
| rptest.tests.datalake.compaction_test.CompactionGapsTest.test_translation_no_gaps.cloud_storage_type=CloudStorageType.S3.catalog_type=CatalogType.REST_JDBC | ducktape | https://buildkite.com/redpanda/redpanda/builds/61563#0194d20c-ff98-4ec0-9f56-38befe604032 | FLAKY | 1/2 |
</details>
<details>
<summary>test results on <a href="https://buildkite.com/redpanda/redpanda/builds/61682">build#61682</a></summary>


| test_id | test_kind | job_url | test_status | passed |
|:---------------------------------------------------------------------------------------------------------------------------------------------------------------------|:------------|:------------------------------------------------------------------------------------------|:--------------|:---------|
| rptest.tests.compaction_recovery_test.CompactionRecoveryTest.test_index_recovery | ducktape | https://buildkite.com/redpanda/redpanda/builds/61682#0194dcb6-b3e7-4275-b585-63769e3a91eb | FLAKY | 1/3 |
| rptest.tests.compaction_recovery_test.CompactionRecoveryUpgradeTest.test_index_recovery_after_upgrade | ducktape | https://buildkite.com/redpanda/redpanda/builds/61682#0194dc9a-225b-4365-b41c-e42b927c3e92 | FLAKY | 1/2 |
| rptest.tests.datalake.compaction_test.CompactionGapsTest.test_translation_no_gaps.cloud_storage_type=CloudStorageType.S3.catalog_type=CatalogType.REST_HADOOP | ducktape | https://buildkite.com/redpanda/redpanda/builds/61682#0194dcb6-b3e7-4275-b585-63769e3a91eb | FLAKY | 1/2 |
| rptest.tests.datalake.compaction_test.CompactionGapsTest.test_translation_no_gaps.cloud_storage_type=CloudStorageType.S3.catalog_type=CatalogType.REST_JDBC | ducktape | https://buildkite.com/redpanda/redpanda/builds/61682#0194dcb6-b3e4-449e-a254-c66f8797a6ea | FLAKY | 1/2 |
| rptest.tests.datalake.custom_partitioning_test.DatalakeCustomPartitioningTest.test_basic.cloud_storage_type=CloudStorageType.S3.catalog_type=CatalogType.REST_HADOOP | ducktape | https://buildkite.com/redpanda/redpanda/builds/61682#0194dcb6-b3e5-4007-9090-5e5e97766310 | FLAKY | 1/2 |
| rptest.tests.partition_movement_test.PartitionMovementTest.test_availability_when_one_node_down | ducktape | https://buildkite.com/redpanda/redpanda/builds/61682#0194dc9a-225a-472c-827d-daaa26f07098 | FLAKY | 1/2 |
| rptest.tests.scaling_up_test.ScalingUpTest.test_scaling_up_with_recovered_topic | ducktape | https://buildkite.com/redpanda/redpanda/builds/61682#0194dcb6-b3e6-4ddf-b17d-2a71ef0b0f40 | FLAKY | 1/2 |
| rptest.tests.write_caching_fi_test.WriteCachingFailureInjectionTest.test_crash_all | ducktape | https://buildkite.com/redpanda/redpanda/builds/61682#0194dc9a-225b-4ae0-9014-9e69b7cda65e | FLAKY | 1/2 |
</details>
<details>
<summary>test results on <a href="https://buildkite.com/redpanda/redpanda/builds/61910">build#61910</a></summary>


| test_id                                                                                                                                                           | test_kind   | job_url                                                                                   | test_status   | passed   |
|:------------------------------------------------------------------------------------------------------------------------------------------------------------------|:------------|:------------------------------------------------------------------------------------------|:--------------|:---------|
| kafka_server_rpfixture.kafka_server_rpfixture                                                                                                                     | unit        | https://buildkite.com/redpanda/redpanda/builds/61910#019513ab-cd50-428f-9d81-5f8116eaf3f3 | FLAKY         | 1/2      |
| rptest.tests.compaction_recovery_test.CompactionRecoveryTest.test_index_recovery                                                                                  | ducktape    | https://buildkite.com/redpanda/redpanda/builds/61910#019513f3-f79a-4ef5-b87b-40ddaa1e0373 | FLAKY         | 1/2      |
| rptest.tests.datalake.datalake_e2e_test.DatalakeE2ETests.test_topic_lifecycle.cloud_storage_type=CloudStorageType.S3.catalog_type=CatalogType.REST_HADOOP         | ducktape    | https://buildkite.com/redpanda/redpanda/builds/61910#01951406-cc75-4968-824a-19b89e830bdd | FLAKY         | 1/2      |
| rptest.tests.datalake.mount_unmount_test.MountUnmountIcebergTest.test_simple_remount.cloud_storage_type=CloudStorageType.S3                                       | ducktape    | https://buildkite.com/redpanda/redpanda/builds/61910#01951406-cc74-4052-9b53-57304652913d | FLAKY         | 1/2      |
| rptest.tests.log_compaction_test.LogCompactionTest.compaction_stress_test.cleanup_policy=compact.key_set_cardinality=1000.storage_compaction_key_map_memory_kb=10 | ducktape    | https://buildkite.com/redpanda/redpanda/builds/61910#019513f3-f798-4e04-a267-6ece0682e028 | FLAKY         | 1/2      |
| rptest.tests.log_compaction_test.LogCompactionTest.compaction_stress_test.cleanup_policy=compact.key_set_cardinality=1000.storage_compaction_key_map_memory_kb=3  | ducktape    | https://buildkite.com/redpanda/redpanda/builds/61910#019513f3-f79a-4ef5-b87b-40ddaa1e0373 | FLAKY         | 1/2      |
| rptest.tests.partition_movement_test.SIPartitionMovementTest.test_shadow_indexing.num_to_upgrade=0.cloud_storage_type=CloudStorageType.ABS                        | ducktape    | https://buildkite.com/redpanda/redpanda/builds/61910#019513f3-f79a-4ef5-b87b-40ddaa1e0373 | FLAKY         | 1/2      |
</details>
<!--/body-->

mmaslankaprv committed:
r/tests: made recovery read size configurable in tests

Signed-off-by: Michał Maślanka <michal@redpanda.com>

mmaslankaprv committed:
r/tests: fixed waiting for offsets in tests

Assertion triggered in function body are not propagated to the test
itself. Change the method to throw an exception in case of timeout
instead of using an assertion.

Signed-off-by: Michał Maślanka <michal@redpanda.com>

mmaslankaprv committed:
r/tests: added ability to set reply interceptor function

The reply interceptor allows test creator to modify or drop the reply
that is about to be processed by the RPC requester. This allow tests to
take more control over the Raft protocol behavior and test some rare
edge cases which might be hard to trigger otherwise.

Signed-off-by: Michał Maślanka <michal@redpanda.com>

mmaslankaprv committed:
r/tests: added test validating processing all matching batches

Signed-off-by: Michał Maślanka <michal@redpanda.com>

bashtanov commented on src/v/raft/tests/raft_fixture.h:
```diff
@@ -272,13 +272,18 @@ class raft_node_instance : public ss::weakly_referencable<raft_node_instance> {
     ss::sstring _base_directory;
     config::mock_property<size_t> _max_inflight_requests{16};
     config::mock_property<size_t> _max_queued_bytes{1_MiB};
+    config::mock_property<size_t> _default_recovery_read_size{32_KiB};
```
any reason we change it for existing tests?

mmaslankaprv replied:
no particular reason, i will make sure it is the same as before

bashtanov commented on src/v/raft/tests/basic_raft_fixture_test.cc:
```diff
@@ -862,3 +862,101 @@ TEST_F_CORO(raft_fixture, test_no_stepdown_on_append_entries_timeout) {
+    ASSERT_FALSE_CORO(r.has_error());
+
+    /**
+     * Append some batches to the blocked node with different term
```
How can this happen naturally? This node's log is shorter than the majority, so it cannot be elected as a leader.

In my test data was produced with acks=1, isn't it what we need here to keep the scenario plausible?

mmaslankaprv replied:
for a short while when a node is still a leader but the others doesn't receive messages it may always happen, using `acks=1` makes it more plausible

bashtanov replied:
Could you check my understanding of how we can make it happen without manual log append?
1. Elect node `not_leader` as a leader in term 1
2. Produce, do not wait for replicated, make sure (by blocking dispatch or append_entries?) it does NOT get replicated to `leader` or `third`
3. Elect node `leader` as a leader in term 2 (we'll need to make node `third` communicated well with `leader` for it)
4. Block off append entries on `not_leader`
5. Produce again, make sure it gets replicated to `third` and thus committed (is it vital for the test?)
6. Proceed with the rest of the test.

The terms would be the other way round (1 on `leader`, 2 on `not_leader`), I'm not sure how to make them the right way round.

Is that right? Do we not do this just to keep the test simple?

mmaslankaprv replied:
exactly, the description here is valid, i just wanted to make it a bit simpler

bashtanov replied:
Could you swap round term values, so that the situation looks more realistic? Or explain how term 2 data can be appended to `non_leader`'s log.

mmaslankaprv replied:
i added a comment and made it more clear with terms

bashtanov commented on src/v/raft/tests/basic_raft_fixture_test.cc:
```diff
@@ -862,3 +862,101 @@ TEST_F_CORO(raft_fixture, test_no_stepdown_on_append_entries_timeout) {
+      [&, term_1_match_offset](reply_variant reply, model::node_id) {
+          return ss::visit(
+            std::move(reply),
+            [&, term_1_match_offset](append_entries_reply& a_r) {
```
TIL `ss::visit` does not forward the value argument to `std::visit`, but silently makes a copy :facepalm:

bashtanov replied:
[It's not going to last long though](https://github.com/scylladb/seastar/pull/2636)

bashtanov commented:
> Assertion triggered in function body are not propagated to the test itself

Why is that? Anything wrong with the macro? AFAIK it's meant to work with both gtest and boost.

bharathv commented on src/v/raft/tests/raft_fixture.cc:
```diff
@@ -226,6 +226,10 @@ void in_memory_test_protocol::on_dispatch(dispatch_callback_t f) {
     _on_dispatch_handlers.push_back(std::move(f));
 }
 
+void in_memory_test_protocol::reset_on_dispatch() {
```
🐎  I did something similar yesterday https://github.com/redpanda-data/redpanda/pull/25035/commits/639ee9ce210e0d221230fde13da287a1bc01c588#diff-fbb5ba4f4b2361c193c53e1a1d2b54740ff65a120b253006addfd00b2c4fc3bbR146

bharathv reviewed (commented):
lgtm modulo one question, took me a bit to digest the change, had to dig up Alexey's change that added these checks. Would be nice to get a blessing from @ztlpn  too.

bashtanov commented on src/v/raft/tests/basic_raft_fixture_test.cc:
```diff
@@ -862,3 +862,108 @@ TEST_F_CORO(raft_fixture, test_no_stepdown_on_append_entries_timeout) {
+     * Append some batches to the blocked node with different term. This is a
+     * 'brute force' solution to generate a situation in which the follower
+     * and leader log would diverge. This simulates a situation in which the
+     * blocked follower replica was a leader in term 2 and accepted some writes
```
> This simulates a situation in which the blocked follower replica was a leader in term 2

... which it could not be elected for. If another node (term 1 leader or third node) voted for it this voter would have bumped its term to 2 when having an empty log. It would not be able to save term 1 data then.

bashtanov commented on src/v/raft/tests/raft_fixture.cc:
```diff
@@ -701,6 +701,22 @@ raft_fixture::wait_for_leader(model::timeout_clock::time_point deadline) {
+      _nodes | std::views::keys
+        | std::views::filter(
+          [leader_id](model::node_id id) { return id != leader_id; }),
+      std::back_inserter(followers));
```
nit: use `copy_if`?

bashtanov commented on src/v/raft/tests/basic_raft_fixture_test.cc:
```diff
@@ -862,3 +862,109 @@ TEST_F_CORO(raft_fixture, test_no_stepdown_on_append_entries_timeout) {
+    /**
+     * Recover communication and wait for the intercept to trigger
+     */
+    new_leader_node.reset_dispatch_handlers();
```
This will enable the new leader to send vote requests to the old leader. I guess it won't anyway, as it has been elected already. Do we need this?

bashtanov commented on src/v/raft/tests/raft_fixture.h:
```diff
@@ -395,7 +395,8 @@ class raft_fixture
       size_t batch_record_count,
-      size_t record_payload_size) {
+      size_t record_payload_size,
+      model::term_id term = model::term_id(0)) {
```
did you decide to keep it just in case it is needed in future?

bashtanov commented on src/v/raft/tests/basic_raft_fixture_test.cc:
```diff
@@ -862,3 +862,109 @@ TEST_F_CORO(raft_fixture, test_no_stepdown_on_append_entries_timeout) {
+     * Recover communication and wait for the intercept to trigger
+     */
+    new_leader_node.reset_dispatch_handlers();
+    co_await reply_intercepted.wait([&] { return intercept_count > 5; });
```
We don't produce anything after the second election. What are the 20+ messages that are replicated from the new leader to the old one?

mmaslankaprv replied:
there will be no new messages replicated, only recovery append entry requests finally leading to truncation

bashtanov reviewed (commented):
A few questions as I'm not sure I understand the test.

vbotbuildovich commented:
/backport v24.3.x

vbotbuildovich commented:
/backport v24.2.x

vbotbuildovich commented:
/backport v24.1.x

vbotbuildovich commented:
Failed to create a backport PR to v24.3.x branch. I tried:

```
git remote add upstream https://github.com/redpanda-data/redpanda.git
git fetch --all
git checkout -b backport-pr-25018-v24.3.x-754 remotes/upstream/v24.3.x
git cherry-pick -x 0d029cfd84 edd55dcf23 bd47cd93a2 e2ec0352df 9627077942 59e43a7c31
```

[Workflow run logs](https://github.com/redpanda-data/redpanda/actions/runs/13394777792).

vbotbuildovich commented:
Failed to create a backport PR to v24.2.x branch. I tried:

```
git remote add upstream https://github.com/redpanda-data/redpanda.git
git fetch --all
git checkout -b backport-pr-25018-v24.2.x-107 remotes/upstream/v24.2.x
git cherry-pick -x 0d029cfd84 edd55dcf23 bd47cd93a2 e2ec0352df 9627077942 59e43a7c31
```

[Workflow run logs](https://github.com/redpanda-data/redpanda/actions/runs/13394777790).

vbotbuildovich commented:
Failed to create a backport PR to v24.1.x branch. I tried:

```
git remote add upstream https://github.com/redpanda-data/redpanda.git
git fetch --all
git checkout -b backport-pr-25018-v24.1.x-534 remotes/upstream/v24.1.x
git cherry-pick -x 0d029cfd84 edd55dcf23 bd47cd93a2 e2ec0352df 9627077942 59e43a7c31
```

[Workflow run logs](https://github.com/redpanda-data/redpanda/actions/runs/13394778557).
